### PR TITLE
Memory efficiency, performance improvements,and architecture refactoring 

### PR DESCRIPTION
**Pull request by adrientoupet**
Hi everyone,

This PR is focused on performance and optimization. SeedVR2 should now be faster while allowing you to upscale at higher resolutions with less memory consumption. I'll move on to support tickets and feature requests after this, as I know there's a backlog. Anyway, give it a shot and let me know if I broke anything (hopefully not too much!)!

Changelog below - refer to individual commits if you want more details, or ask me :)

#### Model Initialization & Loading
- **Faster model initialization** using meta device to avoid unnecessary memory allocation during DiT and VAE creation
- **Memory-mapped VAE loading** with `assign=True` to reference tensors directly instead of copying, matching DiT's loading strategy
- **Optimized state dict management** with explicit deletion after loading to free memory immediately
- **Removed slow preinitialize_rope_cache()** bottleneck during model preparation

#### Generation Pipeline  
- **Streamlined batch processing** by merging redundant log steps and unifying "Model Configuration" and "Input Preparation" into single "Generation Setup" step
- **Optimized memory usage** by passing compute_dtype through entire diffusion pipeline, ensuring timesteps use bfloat16 instead of float32
- **Code deduplication** by reusing existing functions: `prepare_video_transforms()`, `load_text_embeddings()`, `calculate_optimal_batch_params()`
- **Improved batch recommendations** with `calculate_optimal_batch_params()` now prioritizing temporal stability by suggesting largest valid 4n+1 batch size

#### VAE Operations
- **Optimized tile blending** with pre-computed cosine ramps instead of recalculating per-tile
- **In-place operations** using `mul_` and `div_` to reduce memory allocations
- **Reduced redundant transfers** by skipping device moves when tensors are already on correct device
- **Improved slicing operations** with consistent memory cleanup for both encode and decode paths

#### Critical Fixes
- **Fixed major VRAM leak** in VAE decoder by clearing InflatedCausalConv3d memory buffers when moving VAE to CPU
- **Fixed RoPE OOM issues** by computing only required dimensions instead of maximum (1024x128x128), enabling higher resolution upscaling
- **Added OOM retry mechanism** for VAE operations (Upsample3D, ResnetBlock3D, InflatedCausalConv3d, GroupNorm) with single retry on failure
- **Added cuBLAS workspace clearing** for better GPU memory management

#### BlockSwap Integration
- **Full preserve_vram compatibility** with bypass mechanism to temporarily disable BlockSwap protection during offload
- **Transparent handling** in `manage_model_device()` for BlockSwap models
- **Proper state restoration** of blocks to correct devices (partial GPU config) on reload
- **Simplified preserve_vram logic** by removing BlockSwap checks from infer.py

#### Code Organization
- **Unified model configuration** by merging `configure_dit_model_inference` and `configure_vae_model_inference` into single `configure_model_inference` function  
- **Focused helper functions** extracted from `configure_model_inference()` 
- **Removed redundant code**: gradient checkpointing (training-only), model_weight variables, platform-specific VAE dtype logic
- **Simplified FP8 logic** by removing unnecessary conversion code from `load_quantized_state_dict`

#### Dtype Handling
- **Unified dtype flow**: VAE now uses configured dtype consistently, removing redundant vae_dtype variable
- **Fixed FP16 black frames** by using BFloat16 compute/autocast for FP16 models
- **Natural dtype propagation** from generation_loop → configure_diffusion → sampling, eliminating manual conversions
- **Removed nested autocast contexts** relying on outer context from generation_step

#### Debug & Logging
- **Optimized debug performance** by making tensor counting opt-in (`show_tensors=False` by default)
- **Conditional logging** with `reset_vram_peak()` only logging when debug is enabled
- **Early return pattern** in log() method to avoid string operations when disabled
- **Standardized formatting**: consistent .2f float precision, uppercase device names, unified category names
- **Improved clarity**: better frame padding logs, clearer FP8 RoPE conversion messages

### CODE REVIEW

**Comment by adrientoupet** [comment]
Hi @naxci1 - I'm curious to hear your thoughts on this version - I know you were complaining about performance in the nightly branch compared to the main branch. Is that better now? Thank you

```
cd ComfyUI/custom_nodes
rm -rf ComfyUI-SeedVR2_VideoUpscaler
git clone -b nightly https://github.com/numz/ComfyUI-SeedVR2_VideoUpscaler
pip install -r ComfyUI-SeedVR2_VideoUpscaler/requirements.txt
```

**Comment by naxci1** [comment]
Hi @adrientoupet

Thank you for your answer.

I ran a test. With the old code (3b16, 240p to 480p, batch_size=101, block_to_swap=32), the video I was rendering at 3.65 fps has now increased to 4 fps. VRAM usage has decreased. The VRAM is immediately freed after the render is complete. With the VRAM usage reduced, I now tried block_to_swap=16, and it worked. The previous code was giving an OOM error. I could only run it at 32 fps. Now the speed has increased a bit. It started processing at 4.10 fps.

I lowered the block_to_swap value even further to 8, 4, and 0, and I tested it. It still didn't give an OOM error. This is great. These never worked in the old code; it only worked at 32 fps. With block_to_swap=4, it sped up to 4.21 fps, but it processed at an average of 4.18 fps.

It works with block_to_swap=0, but due to insufficient VRAM (my GPU is an RTX5070 Ti with 16 GB), it uses Shared VRAM, which slows the speed. So I started running it at 4. This way, it doesn't use Shared VRAM, and it uses 15.1 GB of VRAM.

The biggest advantage of the new code is that it doesn't give any OOM errors, and I congratulate you on that.

I also ask you to add the (force=True) code to the comfyui_node.py file, because I have to add it myself every time. I disable debug because I'm not a programmer like you, and I don't understand much of what's written there. I'm just a user. That's why I think there are more people like me, and learning FPS is important for them. Therefore, I think it's better to always display the FPS by default. I would appreciate it if it showed the average FPS when the process is complete. --
comfyui_node.py

self.debug.log(f"Batch {batch_idx} - FPS: {batch_fps:.2f} frames/sec", category="timing", force=True)
---

and it gave this error in the logs

C:\ComfyUI\ComfyUI\custom_nodes\ComfyUI-SeedVR2_VideoUpscaler\src\utils\debug.py:405: FutureWarning: `torch.distributed.reduce_op` is deprecated, please use `torch.distributed.ReduceOp` instead 
elif detailed and isinstance(obj, torch.nn.Module):

**Comment by naxci1** [comment]
I could never run the 7b models with the old code; it would give me an OOM error, but with the new code, it even runs the 7b16 model without any problems. Thank you very much and wish you continued success. I hope the full version will be available soon, and the version number will be added, making it easier to update from the comfyui manager.

Today, I ran hundreds of tests and achieved the highest quality and fastest results with the 7b16 model, using batch_size=197 and block_to_swap=24.

Thank you @adrientoupet and everyone involved in this effort and wish you continued success.

image

**Comment by adrientoupet** [comment]
Nice one, thank you very much for providing this feedback - I'm glad it's finally paying off.
Yes, we're getting very close from doing an official release of this :) It always takes way longer than expecting but I really, really hope it will happen in September. Cheers.
